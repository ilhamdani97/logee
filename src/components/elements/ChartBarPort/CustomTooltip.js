import React from 'react';
import style from './styles.scoped.css';
import PropTypes from 'prop-types';
import { thousand } from '../../../utils/format';


const CustomTooltip = (props) => {
  const{ active, payload,customProps } = props;


  const convertToCurrency = (isCurrency,val) => {
    let currency = isCurrency ? `Rp ${thousand(val)}`: val;

    return currency;
  };
  if (active && payload && payload.length) {

    const captionUpper = payload[0].payload[customProps.titleUpper.value];
    const captionBottom = payload[0].payload[customProps.titleBottom.value];
    const bgUpper = [style[customProps.titleUpper.tooltipColor],style.tooltipRadiusUpper].join(' ');
    const bgBottom = [style[customProps.titleBottom.tooltipColor],style.tooltipRadiusBottom].join(' ');
    const subCapUp = customProps.titleUpper.subCaption;
    const subCapBtm = customProps.titleBottom.subCaption;
    return (
      <div className={style.tooltip}>
        <div className={bgUpper}>
          {`
          ${subCapUp.textPosition === 'left' ? subCapUp.textValue : ''}
          ${convertToCurrency(customProps.titleUpper.currency, captionUpper)} 
          ${subCapUp.textPosition === 'right' ? subCapUp.textValue : ''}
          `}

        </div>
        <div className={bgBottom}>
          {`
          ${subCapBtm.textPosition === 'left' ? subCapBtm.textValue : ''}
          ${convertToCurrency(customProps.titleBottom.currency,  captionBottom)} 
          ${subCapBtm.textPosition === 'right' ? subCapBtm.textValue : ''}`}
        </div>

      </div>
    );
  }

  return null;
};


CustomTooltip.defaultProps = {
  active:false,
  customProps:{},
  payload: [],
};

CustomTooltip.propTypes = {
  active:PropTypes.bool,
  customProps: PropTypes.object,
  payload: PropTypes.array,
};


export default CustomTooltip;
