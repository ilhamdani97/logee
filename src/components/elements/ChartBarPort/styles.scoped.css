.tooltip {
  color: white;
  text-align: center;
  border-radius:5px;
}

.tooltipRadiusUpper{
  -webkit-border-top-left-radius: 5px;
-webkit-border-top-right-radius: 5px;
-moz-border-radius-topleft: 5px;
-moz-border-radius-topright: 5px;
border-top-left-radius: 5px;
border-top-right-radius: 5px;
}

.tooltipRadiusBottom{
  -webkit-border-bottom-right-radius: 5px;
-webkit-border-bottom-left-radius: 5px;
-moz-border-radius-bottomright: 5px;
-moz-border-radius-bottomleft: 5px;
border-bottom-right-radius: 5px;
border-bottom-left-radius: 5px;
}

.primary{
  padding: 8px;
  background-color: var(--primary-color);
}

.secondary{
  padding: 8px;
  background-color: var(--blue-color);
}

.responsiveChart{
  width: 100%;
  height: 300px;
}
