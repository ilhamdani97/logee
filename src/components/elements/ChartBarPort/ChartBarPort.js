import React,{ useEffect, useState } from 'react';
import { BarChart, Bar, XAxis, YAxis, Tooltip, ResponsiveContainer, CartesianGrid } from 'recharts';
import PropTypes from 'prop-types';
import CustomTooltip from './CustomTooltip';
import style from './styles.scoped.css';
import { shortCurrency } from '@utils/format';

const ChartBarPort = ({ data,xLabel, tooltip, barType }) => {

  const [charts, setCharts] = useState(data);
  const [posData, setposData] = useState({});
  const { titleUpper, titleBottom } = tooltip;


  useEffect(() => {
    const newData = data.map((v)=>{

      let yStackAxis = v[titleBottom.value] + v[titleUpper.value];
      return {
        ...v,
        'yAxisUnit':shortCurrency(v[titleBottom.value]),
        'xAxisBar': shortCurrency(v[titleUpper.value]),
        'yAxisBar': shortCurrency(v[titleBottom.value]),
        'yStacked':barType === 'stack' ? shortCurrency(yStackAxis) : '0'
      };
    });

    setCharts(newData);
  }, [data]);

  return (
    <div className={style.responsiveChart}>
      <ResponsiveContainer>
        {barType === 'stack' ?
          <BarChart data={charts}>
            <XAxis dataKey={xLabel} />
            <YAxis dataKey="yStacked" orientation="right"   unit="jt"/>
            <Tooltip
              content={<CustomTooltip customProps={tooltip}/>}
              cursor={{ fill: '#efefef' }}
              position={{ x: posData.x - 50, y: posData.y - 70 }}
            />
            <CartesianGrid stroke="#eee" strokeDasharray="5 0"/>
            <Bar barSize={10} dataKey="yAxisBar" fill="#487A9D" onMouseLeave={() => {
              setposData({});
            }} onMouseOver={(el) => {
              el?setposData(el):setposData({});
            }}

            stackId="stackBar"/>
            <Bar barSize={10} dataKey="xAxisBar" fill="#e8590c" onMouseLeave={() => {
              setposData({});
            }} onMouseOver={(el) => {
              el?setposData(el):setposData({});
            }}

            stackId="stackBar"/>
          </BarChart>
          :
          <BarChart data={charts}>
            <XAxis dataKey={xLabel} />
            <YAxis dataKey="yAxisUnit" orientation="right" unit="jt"/>
            <Tooltip
              content={<CustomTooltip customProps={tooltip}/>}
              cursor={{ fill: '#efefef' }}
              position={{ x: posData.x - 50, y: posData.y - 70 }}
            />
            <CartesianGrid stroke="#eee" strokeDasharray="5 0"/>
            <Bar barSize={10} dataKey="yAxisUnit" fill="#e8590c"
              onMouseLeave={() => {
                setposData({});
              }}

              onMouseOver={(el) => {
                el?setposData(el):setposData({});
              }}/>
          </BarChart>
        }
      </ResponsiveContainer>
    </div>
  );
};


ChartBarPort.defaultProps = {
  barType:'',
  data: [],
  stack:'',
  tooltip:{},
  xLabel:'',
};

ChartBarPort.propTypes = {
  barType: PropTypes.string,
  data: PropTypes.array,
  stack: PropTypes.string,
  tooltip:PropTypes.object,
  xLabel: PropTypes.string,
};


export default ChartBarPort;


