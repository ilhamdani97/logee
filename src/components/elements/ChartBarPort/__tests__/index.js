import React,{ useEffect } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import ChartBarPort from '../ChartBarPort';


const tooltip =  {
  'titleUpper': {
    'currency': false,
    'value': 'totalContainer',
    'subCaption': 'Kontainer',
    'tooltipColor': 'primary'
  },
  'titleBottom': {
    'currency': true,
    'value': 'totalTransaction',
    'subCaption': 'Transaksi',
    'tooltipColor': 'primary'
  }
};

const data = [
  {
    'periodNodeId': 1,
    'periodNode': 'Jan',
    'totalTransaction': 967522920,
    'totalContainer': 1,
    'yAxisUnit': 967,
    'xAxisBar': 0,
    'yAxisBar': 967,
    'yStacked': '0'
  }
];

const barType = 'stack';


describe('src/components/elements/ChartBarPort', () => {
  test('render', () => {
    useEffect.mockImplementationOnce(fn => fn());
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<ChartBarPort data={data} tooltip={tooltip}/>);
    expect(tree).toMatchSnapshot();

    const barAction = tree.props.children.props.children.props.children[4].props;
    barAction.onMouseLeave();
    barAction.onMouseOver();
  });

  test('render bar stacj', () => {
    useEffect.mockImplementationOnce(fn => fn());
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<ChartBarPort barType={barType} data={data} tooltip={tooltip}/>);
    expect(tree).toMatchSnapshot();

    const barAction =tree.props.children.props.children.props.children;
    const barActionAtas = barAction[4].props;
    const barActionBawah = barAction[5].props;

    barActionAtas.onMouseLeave();
    barActionAtas.onMouseOver();
    barActionBawah.onMouseLeave();
    barActionBawah.onMouseOver();
  });
});
