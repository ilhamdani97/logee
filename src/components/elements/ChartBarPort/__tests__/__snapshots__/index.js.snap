// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/elements/ChartBarPort render 1`] = `
<div>
  <ForwardRef>
    <BarChart
      barCategoryGap="10%"
      barGap={4}
      data={
        Array [
          Object {
            "periodNode": "Jan",
            "periodNodeId": 1,
            "totalContainer": 1,
            "totalTransaction": 967522920,
            "xAxisBar": 0,
            "yAxisBar": 967,
            "yAxisUnit": 967,
            "yStacked": "0",
          },
        ]
      }
      layout="horizontal"
      margin={
        Object {
          "bottom": 5,
          "left": 5,
          "right": 5,
          "top": 5,
        }
      }
      reverseStackOrder={false}
      stackOffset="none"
      syncMethod="index"
    >
      <XAxis
        allowDataOverflow={false}
        allowDecimals={true}
        allowDuplicatedCategory={true}
        dataKey=""
        domain={
          Array [
            0,
            "auto",
          ]
        }
        height={30}
        hide={false}
        mirror={false}
        orientation="bottom"
        padding={
          Object {
            "left": 0,
            "right": 0,
          }
        }
        reversed={false}
        scale="auto"
        tickCount={5}
        type="category"
        width={0}
        xAxisId={0}
      />
      <YAxis
        allowDataOverflow={false}
        allowDecimals={true}
        allowDuplicatedCategory={true}
        dataKey="yAxisUnit"
        domain={
          Array [
            0,
            "auto",
          ]
        }
        height={0}
        hide={false}
        mirror={false}
        orientation="right"
        padding={
          Object {
            "bottom": 0,
            "top": 0,
          }
        }
        reversed={false}
        scale="auto"
        tickCount={5}
        type="number"
        unit="jt"
        width={60}
        yAxisId={0}
      />
      <Tooltip
        active={false}
        allowEscapeViewBox={
          Object {
            "x": false,
            "y": false,
          }
        }
        animationDuration={400}
        animationEasing="ease"
        content={
          <CustomTooltip
            active={false}
            customProps={
              Object {
                "titleBottom": Object {
                  "currency": true,
                  "subCaption": "Transaksi",
                  "tooltipColor": "primary",
                  "value": "totalTransaction",
                },
                "titleUpper": Object {
                  "currency": false,
                  "subCaption": "Kontainer",
                  "tooltipColor": "primary",
                  "value": "totalContainer",
                },
              }
            }
            payload={Array []}
          />
        }
        contentStyle={Object {}}
        coordinate={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        cursor={
          Object {
            "fill": "#efefef",
          }
        }
        cursorStyle={Object {}}
        filterNull={true}
        isAnimationActive={true}
        itemStyle={Object {}}
        labelStyle={Object {}}
        offset={10}
        position={
          Object {
            "x": NaN,
            "y": NaN,
          }
        }
        separator=" : "
        trigger="hover"
        useTranslate3d={false}
        viewBox={
          Object {
            "x1": 0,
            "x2": 0,
            "y1": 0,
            "y2": 0,
          }
        }
        wrapperStyle={Object {}}
      />
      <CartesianGrid
        fill="none"
        horizontal={true}
        horizontalFill={Array []}
        horizontalPoints={Array []}
        stroke="#eee"
        strokeDasharray="5 0"
        vertical={true}
        verticalFill={Array []}
        verticalPoints={Array []}
      />
      <Bar
        animationBegin={0}
        animationDuration={400}
        animationEasing="ease"
        barSize={10}
        data={Array []}
        dataKey="yAxisUnit"
        fill="#e8590c"
        hide={false}
        isAnimationActive={true}
        layout="vertical"
        legendType="rect"
        minPointSize={0}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        xAxisId={0}
        yAxisId={0}
      />
    </BarChart>
  </ForwardRef>
</div>
`;

exports[`src/components/elements/ChartBarPort render bar stacj 1`] = `
<div>
  <ForwardRef>
    <BarChart
      barCategoryGap="10%"
      barGap={4}
      data={
        Array [
          Object {
            "periodNode": "Jan",
            "periodNodeId": 1,
            "totalContainer": 1,
            "totalTransaction": 967522920,
            "xAxisBar": 0,
            "yAxisBar": 967,
            "yAxisUnit": 967,
            "yStacked": "0",
          },
        ]
      }
      layout="horizontal"
      margin={
        Object {
          "bottom": 5,
          "left": 5,
          "right": 5,
          "top": 5,
        }
      }
      reverseStackOrder={false}
      stackOffset="none"
      syncMethod="index"
    >
      <XAxis
        allowDataOverflow={false}
        allowDecimals={true}
        allowDuplicatedCategory={true}
        dataKey=""
        domain={
          Array [
            0,
            "auto",
          ]
        }
        height={30}
        hide={false}
        mirror={false}
        orientation="bottom"
        padding={
          Object {
            "left": 0,
            "right": 0,
          }
        }
        reversed={false}
        scale="auto"
        tickCount={5}
        type="category"
        width={0}
        xAxisId={0}
      />
      <YAxis
        allowDataOverflow={false}
        allowDecimals={true}
        allowDuplicatedCategory={true}
        dataKey="yStacked"
        domain={
          Array [
            0,
            "auto",
          ]
        }
        height={0}
        hide={false}
        mirror={false}
        orientation="right"
        padding={
          Object {
            "bottom": 0,
            "top": 0,
          }
        }
        reversed={false}
        scale="auto"
        tickCount={5}
        type="number"
        unit="jt"
        width={60}
        yAxisId={0}
      />
      <Tooltip
        active={false}
        allowEscapeViewBox={
          Object {
            "x": false,
            "y": false,
          }
        }
        animationDuration={400}
        animationEasing="ease"
        content={
          <CustomTooltip
            active={false}
            customProps={
              Object {
                "titleBottom": Object {
                  "currency": true,
                  "subCaption": "Transaksi",
                  "tooltipColor": "primary",
                  "value": "totalTransaction",
                },
                "titleUpper": Object {
                  "currency": false,
                  "subCaption": "Kontainer",
                  "tooltipColor": "primary",
                  "value": "totalContainer",
                },
              }
            }
            payload={Array []}
          />
        }
        contentStyle={Object {}}
        coordinate={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        cursor={
          Object {
            "fill": "#efefef",
          }
        }
        cursorStyle={Object {}}
        filterNull={true}
        isAnimationActive={true}
        itemStyle={Object {}}
        labelStyle={Object {}}
        offset={10}
        position={
          Object {
            "x": NaN,
            "y": NaN,
          }
        }
        separator=" : "
        trigger="hover"
        useTranslate3d={false}
        viewBox={
          Object {
            "x1": 0,
            "x2": 0,
            "y1": 0,
            "y2": 0,
          }
        }
        wrapperStyle={Object {}}
      />
      <CartesianGrid
        fill="none"
        horizontal={true}
        horizontalFill={Array []}
        horizontalPoints={Array []}
        stroke="#eee"
        strokeDasharray="5 0"
        vertical={true}
        verticalFill={Array []}
        verticalPoints={Array []}
      />
      <Bar
        animationBegin={0}
        animationDuration={400}
        animationEasing="ease"
        barSize={10}
        data={Array []}
        dataKey="yAxisBar"
        fill="#487A9D"
        hide={false}
        isAnimationActive={true}
        layout="vertical"
        legendType="rect"
        minPointSize={0}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        stackId="stackBar"
        xAxisId={0}
        yAxisId={0}
      />
      <Bar
        animationBegin={0}
        animationDuration={400}
        animationEasing="ease"
        barSize={10}
        data={Array []}
        dataKey="xAxisBar"
        fill="#e8590c"
        hide={false}
        isAnimationActive={true}
        layout="vertical"
        legendType="rect"
        minPointSize={0}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        stackId="stackBar"
        xAxisId={0}
        yAxisId={0}
      />
    </BarChart>
  </ForwardRef>
</div>
`;
